File: sourcerer-core/src/main/java/org/elder/sourcerer2/DefaultEventSubscriptionFactory.java
package org.elder.sourcerer2;

public class DefaultEventSubscriptionFactory<T> implements EventSubscriptionFactory<T> {
    private final EventRepository<T> repository;

    public DefaultEventSubscriptionFactory(final EventRepository<T> repository) {
        this.repository = repository;
    }

    @Override
    public EventSubscription fromSubscriptionHandler(
            final EventSubscriptionHandler<T> handler,
            final SubscriptionWorkerConfig config,
            final Integer shard
    ) {
        return new DefaultEventSubscription<>(repository, shard, handler, config);
    }
}

File: sourcerer-core/src/main/java/org/elder/sourcerer2/DefaultEventSubscriptionFactory.kt
package org.elder.sourcerer2

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers

class DefaultEventSubscriptionFactory<T> @JvmOverloads constructor(
        private val repository: EventRepository<T>,
        private val dispatcher: CoroutineDispatcher = Dispatchers.IO
) : EventSubscriptionFactory<T> {
    override fun fromSubscriptionHandler(
            handler: EventSubscriptionHandler<T>,
            config: SubscriptionWorkerConfig,
            shard: RepositoryShard?
    ): EventSubscription {
        return DefaultEventSubscription(dispatcher, repository, shard, handler, config)
    }
}
